File: packages/typescript/server/src/sinks.ts
/**
 * Checkpoint sink interface and implementations.
 *
 * Sinks receive signed checkpoints and persist them to external storage
 * (local filesystem, S3, etc.).
 */

import { writeFileSync, mkdirSync } from "fs";
import { join } from "path";

export interface CheckpointSink {
  /** Publish a signed checkpoint (body + detached JWS signature). */
  publish(signedCheckpoint: {
    body: Record<string, unknown>;
    signature: string;
  }): void;
}

/**
 * Writes signed checkpoints as JSON files to a local directory.
 */
export class LocalFileSink implements CheckpointSink {
  private _directory: string;

  constructor(directory: string) {
    this._directory = directory;
    mkdirSync(directory, { recursive: true });
  }

  publish(signedCheckpoint: {
    body: Record<string, unknown>;
    signature: string;
  }): void {
    const body = signedCheckpoint.body;
    const filename = `${body.checkpoint_id as string}.json`;
    const path = join(this._directory, filename);
    writeFileSync(path, JSON.stringify(signedCheckpoint, null, 2));
  }
}

File: packages/python/anip-server/src/anip_server/sinks.py
"""Checkpoint sink interface and implementations."""
import json
import os
from abc import ABC, abstractmethod
from typing import Any


class CheckpointSink(ABC):
    """Interface for publishing signed checkpoints to external storage."""

    @abstractmethod
    def publish(self, signed_checkpoint: dict[str, Any]) -> None:
        """Publish a signed checkpoint (body + detached JWS signature)."""
        ...


class LocalFileSink(CheckpointSink):
    """Writes signed checkpoints as JSON files to a local directory."""

    def __init__(self, directory: str) -> None:
        self._directory = directory
        os.makedirs(directory, exist_ok=True)

    def publish(self, signed_checkpoint: dict[str, Any]) -> None:
        body = signed_checkpoint["body"]
        filename = f"{body['checkpoint_id']}.json"
        path = os.path.join(self._directory, filename)
        with open(path, "w") as f:
            json.dump(signed_checkpoint, f, indent=2, sort_keys=True)
